vault/ssorolecredentialsprovider.go

```diff
@@ -0,0 +1,257 @@
+	ID         string
+	Secret     string
+	Expiration time.Time
+}
+
+type SSOAccessToken struct {
+	Token      string
+	Expiration time.Time
+}
+
+type SSOOIDCProvider struct {
+	OIDCClient *ssooidc.SSOOIDC
+	Keyring    *CredentialKeyring
+	StartURL   string
+}
+
+func (p *SSOOIDCProvider) GetAccessToken() (*SSOAccessToken, error) {
+	var (
+		creds = &struct {
+			Token  *SSOAccessToken
+			Client *SSOClientCredentials
+		}{
+			Client: &SSOClientCredentials{},
+			Token:  &SSOAccessToken{},
+		}
+		credsUpdated bool
+	)
+
+	item, err := p.Keyring.Keyring.Get(p.StartURL)
+	if err != nil && err != keyring.ErrKeyNotFound {
+		return nil, err
+	}
+
+	if item.Data != nil {
+		if err = json.Unmarshal(item.Data, &creds); err != nil {
+			return nil, fmt.Errorf("Invalid data in keyring: %v", err)
+		}
+	}
+
+	if creds.Client.Expiration.Before(time.Now()) {
+		creds.Client, err = p.registerNewClient()
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO client for %s (expires at: %s)", p.StartURL, creds.Client.Expiration.String())
+		credsUpdated = true
+	}
+
+	if creds.Token.Expiration.Before(time.Now()) {
+		creds.Token, err = p.createClientToken(creds.Client)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO access token for %s (expires at: %s)", p.StartURL, creds.Token.Expiration.String())
+		credsUpdated = true
+	}
+
+	if credsUpdated {
+		bytes, err := json.Marshal(creds)
+		if err != nil {
+			return nil, err
+		}
+		err = p.Keyring.Keyring.Set(keyring.Item{
+			Key:                         p.StartURL,
+			Label:                       fmt.Sprintf("aws-vault (%s)", p.StartURL),
+			Data:                        bytes,
+			KeychainNotTrustApplication: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return creds.Token, nil
+}
+
+func (p *SSOOIDCProvider) registerNewClient() (*SSOClientCredentials, error) {
+	c, err := p.OIDCClient.RegisterClient(&ssooidc.RegisterClientInput{
+		ClientName: aws.String(ssoClientName),
+		ClientType: aws.String(ssoClientType),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &SSOClientCredentials{
+		ID:         aws.StringValue(c.ClientId),
+		Secret:     aws.StringValue(c.ClientSecret),
+		Expiration: time.Unix(aws.Int64Value(c.ClientSecretExpiresAt), 0),
+	}, nil
+}
+
+func (p *SSOOIDCProvider) createClientToken(creds *SSOClientCredentials) (*SSOAccessToken, error) {
+	auth, err := p.OIDCClient.StartDeviceAuthorization(&ssooidc.StartDeviceAuthorizationInput{
+		ClientId:     aws.String(creds.ID),
+		ClientSecret: aws.String(creds.Secret),
+		StartUrl:     aws.String(p.StartURL),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf(authorizationTemplate, aws.StringValue(auth.VerificationUriComplete))
+
+	if err := open.Run(aws.StringValue(auth.VerificationUriComplete)); err != nil {
+		log.Printf("failed to open browser: %s", err)
+	}
+
+	for {
+		// Sleep to allow the user to complete the login flow
+		time.Sleep(3 * time.Second)
```

Comment:
I wonder about this delay, should it be configurable? I am not very fast on the keyboard, it could definitely take me longer than 3 seconds to enter my credentials in AWS SSO via my 3rd party IdP, including finding my MFA device and completing the authentication. Or is it that the user should already be logged into the browser ahead of time and this time is for clicking a single approve button?

Reply:
Just to clarify, retries every 3 seconds. If you have a SSO session against your IDP, it doesn’t require you to re-authenticate with the MFA, so waiting longer might be annoying as it’s almost instant after the first initial authentication.

Reply:
I did some digging around to see how the AWS CLI does this, and ended up here: https://github.com/boto/botocore/blob/v2/botocore/utils.py#L1663

And from their comments around the default retry interval and slow down delay:
https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.5

I.e., we (like the AWS CLI) should just be using the interval that is specified in the response from the API when calling `StartDeviceAuthorization`, and fall back to the 5s defaults defined by the RFC. https://github.com/99designs/aws-vault/pull/549/commits/d7ce500b1aed44ed9d4d2d3dbe649c035ddaa747 implements this, and also adds handling for the `SlowDownException` error code.

From my testing, this has effectively reduced the retry interval to 1s, which makes AWS Vault more snappy to respond after I had completed the browser login.